core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/refresh/extension/InteractionRefreshExtension.java

```diff
@@ -0,0 +1,188 @@
+/*******************************************************************************
+ * Copyright (c) 2020 THALES GLOBAL SERVICES.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Contributors:
+ *    Thales - initial API and implementation
+ *******************************************************************************/
+package org.polarsys.capella.core.sirius.analysis.refresh.extension;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.sirius.diagram.DDiagram;
+import org.eclipse.sirius.diagram.business.api.refresh.IRefreshExtension;
+import org.polarsys.capella.core.data.helpers.cache.Cache;
+import org.polarsys.capella.core.data.interaction.InstanceRole;
+import org.polarsys.capella.core.data.interaction.InteractionFragment;
+import org.polarsys.capella.core.data.interaction.TimeLapse;
+import org.polarsys.capella.core.sirius.analysis.EventContextServices.EventContext;
+
+public class InteractionRefreshExtension implements IRefreshExtension {
+
+  /**
+   * Generic cache (basically for getCovered).
+   */
+  private static Cache interactionCache = new Cache();
+
+  /**
+   * Cache for MessageEnd -> Execution/InstanceRole and InteractionState -> StateFragment. (Cross referencer might be a
+   * way to get this information.)
+   */
+  private static Map<InteractionFragment, TimeLapse> interactionFragmentToTimeLapseCache = new ConcurrentHashMap<>();
+
+  /**
+   * Cache for ChildExecution -> ParentExecution and StateFragment -> ParentExecution.
+   */
+  private static Map<EObject, Collection<EObject>> directEventsCache = new ConcurrentHashMap<>();
+
+  /**
+   * Cache for InstanceRole -> EventContext structure.
+   */
+  private static Map<InstanceRole, List<EventContext>> instanceRoleToEventContextsCache = new ConcurrentHashMap<>();
+
+  /**
+   * Is InteractionRefreshExtension caches enabled ?
+   */
+  private static boolean isCacheEnabled;
+
+  /**
+   * @see org.eclipse.sirius.business.api.refresh.IRefreshExtension#beforeRefresh(org.eclipse.sirius.DDiagram)
+   */
+  @Override
+  public void beforeRefresh(DDiagram diagram) {
+      enableRefreshCache();
```

Comment:
We should also call `clearCache()` here, in case the root `postRefresh` Sirius method fails for whatever reason, and our `postRefresh` is not called.